### ISSUE TITLE
Do we need the concept of ValueResovlerContexts?

### ISSUE THREAD

**Issue by mattbrailsford** [leekelleher/umbraco-ditto#94]
So in my DitFlo demo, I've been making use of ValueResovlers a lot to resovle view model properties. There comes a time though when the value depends on certain flags within the request. Now we could have the value resovler just access the request, or we can DittoIgnore the property and we construct the value ourselves within a controller, however I was wondering if the concept of a ValueResovlerContext, that can be set at runtime, could be a handy feature.

In the DitFlo demo we have a news listing page that lists news items in a paged manor. To get the news items I'm using a value resovler, but obviously, the value resolver needs to know the current page so that it can grab the right news items. 

This is my current implementation which directly accesses the HttpContext Request object  is as follows:

    public class NewsResolver : BaseNewsResolver
    {
        public override object ResolveValue()
        {
            var currentPage = long.Parse("0" + HttpContext.Current.Request["p"]);
            var pageSize = 2;

            var items = GetNews().ToList();
            var totalItems = items.Count;
            var totalPages = (long)Math.Ceiling(totalItems/(decimal)pageSize);

            currentPage = Math.Max(1, Math.Min(currentPage, totalPages));

            var pagedItems = items.Skip((int)(currentPage-1)*pageSize).Take(pageSize)
                .As<NewsPreviewLink>();

            return new PagedCollection<NewsPreviewLink>
            {
                CurrentPage = currentPage,
                PageSize = pageSize,
                TotalItems = totalItems,
                TotalPages = totalPages,
                Items = pagedItems
            };
        }
    }

However from discussions held here https://github.com/mattbrailsford/umbraco-ditflo/issues/1#issuecomment-113080492 I've come up with a possible solution using ValueResolverContexts that could make that as follows

**Controller**  

    public class NewsOverviewController : DitFloController {
        public ActionResult Index(RenderModel model, int p) {
            Ditto.RegisterResolverContext(new NewsResolverContext{
                CurrentPage = p
            };
            return CurrentView();
        }
    }

**Resolver**  

    public class NewsResolver : BaseNewsResolver
    {
        private NewsResolverContext _context;

        public NewsResolver()
            : this(new NewsResolverContext())
        { }

        public NewsResolver(NewsResolverContext ctx)
        {
            _context = ctx;
        }

        public override object ResolveValue()
        {
            var currentPage = _context.CurrentPage;
            var pageSize = 2;

            var items = GetNews().ToList();
            var totalItems = items.Count;
            var totalPages = (long)Math.Ceiling(totalItems/(decimal)pageSize);

            currentPage = Math.Max(1, Math.Min(currentPage, totalPages));

            var pagedItems = items.Skip((int)(currentPage-1)*pageSize).Take(pageSize)
                .As<NewsPreviewLink>();

            return new PagedCollection<NewsPreviewLink>
            {
                CurrentPage = currentPage,
                PageSize = pageSize,
                TotalItems = totalItems,
                TotalPages = totalPages,
                Items = pagedItems
            };
        }
    }

    public class NewsResolverContext : DittoValueResovlerContext
    {
        public int CurrentPage { get; set; }

        public NewsResolverContext()
        {
            CurrentPage = 1;
        }
    }

The idea being that when Ditto resolves the value, it looks for a constructor on the resolver that accepts a context, and if it finds one, look to see if an exact type match has been registered during the current request and if it has, construct the resolver passing in the context.

The main purpose of this is basically decouple the resolver from having to know where to fetch these values from. The controller looks them up and stores them, then ditto handles creating the model and passing them,

Thoughts?

**Comment by mattbrailsford** [comment]
1 thing on my mind, does this conflict with Resolver Attributes at all? or do we say attributes are for fixed config, and contexts are for runtime config?

**Comment by mattbrailsford** [comment]
I think if we go this route, we may also need to change some terminology as we currently call the TypeDescriptorContext "context" too, so we may need a way to explicitly distinguish the purpose of the two.

**Comment by mattbrailsford** [comment]
So in my latest commit 4f1282902e934116db80eeffab6b7bd85608f8f2 I've made a bunch of changes that now makes the resolver context the same as the type descriptor context so there is now just a single "context" concept within a resolver. Now it's a case of, if someone has already registered a context, we used that and populate the ITypeDescriptor properties of that object, rather than creating a new one. I think this feels much better but I'd really like a second opinion on this stuff.
